src/library/scala/collection/immutable/TreeMap.scala

```diff
@@ -157,12 +171,20 @@ object TreeMap extends SortedMapFactory[TreeMap] {
   def from[K : Ordering, V](it: IterableOnce[(K, V)]): TreeMap[K, V] =
     it match {
       case tm: TreeMap[K, V] => tm
-      case _ => (newBuilder[K, V] ++= it).result()
+      case _ =>
+        var t: RB.Tree[K, V] = null
+        val i = it.iterator
+        while (i.hasNext) {
+          val (k, v) = i.next()
+          t = RB.update(t, k, v, overwrite = true)
+        }
+        new TreeMap[K, V](t)
     }
 
   def newBuilder[K : Ordering, V]: Builder[(K, V), TreeMap[K, V]] =
     new ImmutableBuilder[(K, V), TreeMap[K, V]](empty) {
       def addOne(elem: (K, V)): this.type = { elems = elems + elem; this }
+      override def addAll(xs: IterableOnce[(K, V)]): this.type = { elems = elems.concat(xs); this }
```

Comment:
We could go down the path of `HashMapBuilder` and defer creation of the collection that wraps the `RBTree` until result, and make `addOne` just call `t = RB.update(t, k, v, overwrite = true)` like you've done above.

Going further, we could even use mutation of the trees during building. 

/cc @joshlemer

Reply:
Do we have a good solution for the memory barrier problem in `List` and `Vector` yet? Switching to a mutable tree would get us into the same situation here.